Sparse CSR CUDA: Add torch.sparse.sampled_addmm

IvanYashchuk opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #68084
* #68083
* __->__ #68007

This PR adds a new function to the sparse module.
`sampled_addmm` computes α*(A @ B) * spy(C) + β*C, where C is a sparse CSR matrix and A, B are dense (strided) matrices.
This function is currently restricted to single 2D matrices, it doesn't support batched input.

cc @nikitaved @pearu @cpuhrsch @IvanYashchuk

Differential Revision: [D32435799](https://our.internmc.facebook.com/intern/diff/D32435799)

IvanYashchuk committed:
[Sparse CSR CUDA] Add torch.sparse.sampled_addmm

This PR adds a new function to the sparse module.
`sampled_addmm` computes α*(A @ B) * spy(C) + β*C, where C is a sparse CSR matrix and A, B are dense (strided) matrices.
This function is currently restricted to single 2D matrices, it doesn't
support batched input.

[ghstack-poisoned]

cpuhrsch commented on torch/sparse/__init__.py:
```diff
@@ -91,6 +90,20 @@ def mm(mat1: Tensor, mat2: Tensor) -> Tensor:
+sampled_addmm = _add_docstr(_sparse.sampled_addmm, r"""
+sparse.sampled_addmm(input, mat1, mat2, *, beta=1., alpha=1., out=None) -> Tensor
+
+Computes alpha*(mat1 @ mat2) * spy(input) + beta*input, where spy(input) is the sparsity pattern matrix of input.
```
nit: I'd also define what "sparsity pattern" means here. and that it's constrained to CSR

IvanYashchuk replied:
I updated the documentation.

cpuhrsch commented on aten/src/ATen/native/sparse/cuda/SparseBlasImpl.cpp:
```diff
@@ -786,6 +786,85 @@ void triangular_solve_out_sparse_csr(
+  auto descB = at::cuda::sparse::CuSparseDnMatDescriptor(B);
+  auto descC = at::cuda::sparse::CuSparseSpMatCsrDescriptor(C);
+
+  cusparseOperation_t opA = CUSPARSE_OPERATION_NON_TRANSPOSE;
```
Does this mean the operands need to be contiguous? I don't think you're checking that.

cpuhrsch reviewed (changes requested):
Looks good overall, added some small comments.

IvanYashchuk committed:
Update on "[Sparse CSR CUDA] Add torch.sparse.sampled_addmm"


This PR adds a new function to the sparse module.
`sampled_addmm` computes α*(A @ B) * spy(C) + β*C, where C is a sparse CSR matrix and A, B are dense (strided) matrices.
This function is currently restricted to single 2D matrices, it doesn't support batched input.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "[Sparse CSR CUDA] Add torch.sparse.sampled_addmm"


This PR adds a new function to the sparse module.
`sampled_addmm` computes α*(A @ B) * spy(C) + β*C, where C is a sparse CSR matrix and A, B are dense (strided) matrices.
This function is currently restricted to single 2D matrices, it doesn't support batched input.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "[Sparse CSR CUDA] Add torch.sparse.sampled_addmm"


This PR adds a new function to the sparse module.
`sampled_addmm` computes α*(A @ B) * spy(C) + β*C, where C is a sparse CSR matrix and A, B are dense (strided) matrices.
This function is currently restricted to single 2D matrices, it doesn't support batched input.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: Add torch.sparse.sampled_addmm"


This PR adds a new function to the sparse module.
`sampled_addmm` computes α*(A @ B) * spy(C) + β*C, where C is a sparse CSR matrix and A, B are dense (strided) matrices.
This function is currently restricted to single 2D matrices, it doesn't support batched input.

cc nikitaved pearu cpuhrsch IvanYashchuk

[ghstack-poisoned]

cpuhrsch reviewed (approved):
Great! Thanks!

Should we add a very basic (and slow) CPU fallback just for usability / debugging in a separate PR?

Could be as simple as densifying the sparse input Tensor and executing the reference implementation from the tests.

IvanYashchuk committed:
Update on "Sparse CSR CUDA: Add torch.sparse.sampled_addmm"


This PR adds a new function to the sparse module.
`sampled_addmm` computes α*(A @ B) * spy(C) + β*C, where C is a sparse CSR matrix and A, B are dense (strided) matrices.
This function is currently restricted to single 2D matrices, it doesn't support batched input.

cc nikitaved pearu cpuhrsch IvanYashchuk

[ghstack-poisoned]

IvanYashchuk commented:
> Should we add a very basic (and slow) CPU fallback just for usability / debugging in a separate PR?
> Could be as simple as densifying the sparse input Tensor and executing the reference implementation from the tests.

We can add it once `torch.ones_like` works for sparse matrices.

I reordered the PRs as requested, placing https://github.com/pytorch/pytorch/pull/68006 after this one.

cpuhrsch commented:
@cpuhrsch has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32435799).

cpuhrsch commented:
@IvanYashchuk - can you take a look into the test failures? You might need to rebase

IvanYashchuk commented:
The next PR in the stack doesn't fail these tests, that's why a `python_module: sparse` is needed 🙂 . I will fix the failures.

ROCm build failure is real as well.

IvanYashchuk committed:
Update on "Sparse CSR CUDA: Add torch.sparse.sampled_addmm"


This PR adds a new function to the sparse module.
`sampled_addmm` computes α*(A @ B) * spy(C) + β*C, where C is a sparse CSR matrix and A, B are dense (strided) matrices.
This function is currently restricted to single 2D matrices, it doesn't support batched input.

cc nikitaved pearu cpuhrsch IvanYashchuk

Differential Revision: [D32435799](https://our.internmc.facebook.com/intern/diff/D32435799)

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: Add torch.sparse.sampled_addmm"


This PR adds a new function to the sparse module.
`sampled_addmm` computes α*(A @ B) * spy(C) + β*C, where C is a sparse CSR matrix and A, B are dense (strided) matrices.
This function is currently restricted to single 2D matrices, it doesn't support batched input.

cc nikitaved pearu cpuhrsch IvanYashchuk

Differential Revision: [D32435799](https://our.internmc.facebook.com/intern/diff/D32435799)

[ghstack-poisoned]

cpuhrsch commented:
@cpuhrsch has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32435799).

cpuhrsch commented:
@IvanYashchuk https://github.com/pytorch/pytorch/pull/68672 was merged, please rebase

IvanYashchuk committed:
Update on "Sparse CSR CUDA: Add torch.sparse.sampled_addmm"


This PR adds a new function to the sparse module.
`sampled_addmm` computes α*(A @ B) * spy(C) + β*C, where C is a sparse CSR matrix and A, B are dense (strided) matrices.
This function is currently restricted to single 2D matrices, it doesn't support batched input.

cc nikitaved pearu cpuhrsch IvanYashchuk

Differential Revision: [D32435799](https://our.internmc.facebook.com/intern/diff/D32435799)

[ghstack-poisoned]

IvanYashchuk commented:
Rebased.

cpuhrsch commented:
@cpuhrsch has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32435799).

cpuhrsch commented:
ROCm failure seems legit

```
13:25:38 /var/lib/jenkins/workspace/aten/src/ATen/native/sparse/hip/SparseBlasImpl.cpp:807:34: error: ‘CuSparseDnMatDescriptor’ is not a member of ‘at::cuda::sparse’; did you mean ‘CuSparseMatDescriptor’?
```

It's ok to guard and disable this functionality for AMD

IvanYashchuk committed:
Update on "Sparse CSR CUDA: Add torch.sparse.sampled_addmm"


This PR adds a new function to the sparse module.
`sampled_addmm` computes α*(A @ B) * spy(C) + β*C, where C is a sparse CSR matrix and A, B are dense (strided) matrices.
This function is currently restricted to single 2D matrices, it doesn't support batched input.

cc nikitaved pearu cpuhrsch IvanYashchuk

Differential Revision: [D32435799](https://our.internmc.facebook.com/intern/diff/D32435799)

[ghstack-poisoned]

cpuhrsch commented:
@cpuhrsch has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32435799).

IvanYashchuk committed:
Update on "Sparse CSR CUDA: Add torch.sparse.sampled_addmm"


This PR adds a new function to the sparse module.
`sampled_addmm` computes α*(A @ B) * spy(C) + β*C, where C is a sparse CSR matrix and A, B are dense (strided) matrices.
This function is currently restricted to single 2D matrices, it doesn't support batched input.

cc nikitaved pearu cpuhrsch IvanYashchuk

Differential Revision: [D32435799](https://our.internmc.facebook.com/intern/diff/D32435799)

[ghstack-poisoned]

cpuhrsch commented:
@cpuhrsch has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32435799).

cpuhrsch commented:
This PR fails on Windows internally. Can we guard against it and disable Windows support?

IvanYashchuk commented:
> This PR fails on Windows internally. Can we guard against it and disable Windows support?
> 
> The error is
> 
> ```
> aten\src\aten\native\sparse\cuda\sparseblasimpl.cpp(807): error C2039: 'CuSparseDnMatDescriptor': is not a member of 'at::cuda::sparse'
> ```

Could you please tell what CUDA Toolkit version is used there? GitHub Windows CI passes.

IvanYashchuk commented:
@pytorchbot ciflow rerun

cpuhrsch commented:
The error in [linux-bionic-cuda10.2-py3.9-gcc7 / test (slow, 1, 1, linux.4xlarge.nvidia.gpu)](https://github.com/pytorch/pytorch/runs/4298473976?check_suite_focus=true_) doesn't seem relevant.

cpuhrsch reviewed (changes requested):
Please guard against Windows support for now and create a separate issue that tracks that.

IvanYashchuk committed:
Update on "Sparse CSR CUDA: Add torch.sparse.sampled_addmm"


This PR adds a new function to the sparse module.
`sampled_addmm` computes α*(A @ B) * spy(C) + β*C, where C is a sparse CSR matrix and A, B are dense (strided) matrices.
This function is currently restricted to single 2D matrices, it doesn't support batched input.

cc nikitaved pearu cpuhrsch IvanYashchuk

Differential Revision: [D32435799](https://our.internmc.facebook.com/intern/diff/D32435799)

[ghstack-poisoned]

IvanYashchuk commented:
> This PR fails on Windows internally. Can we guard against it and disable Windows support?
> 
> The error is
> 
> ```
> aten\src\aten\native\sparse\cuda\sparseblasimpl.cpp(807): error C2039: 'CuSparseDnMatDescriptor': is not a member of 'at::cuda::sparse'
> ```

`CuSparseDnMatDescriptor` was used already in this file and all uses are already guarded by the CUDA Toolkit version.

cpuhrsch commented:
@cpuhrsch has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32435799).

bdhirsh commented on torch/sparse/__init__.py:
```diff
@@ -91,6 +91,57 @@ def mm(mat1: Tensor, mat2: Tensor) -> Tensor:
 
 
+sampled_addmm = _add_docstr(_sparse.sparse_sampled_addmm, r"""
+sparse.sampled_addmm(input, mat1, mat2, *, beta=1., alpha=1., out=None) -> Tensor
```
Hey @IvanYashchuk! It looks like this op is in the public python API for the 1.11 release (`torch.sparse.sampled_addm`), but I don't think these docs made it into the website (I think you need to modify the docs .rst file for that?). Do you mind making a patch and cherry-picking it into the 1.11 release branch?

IvanYashchuk replied:
Right, that was an oversight. I submitted a PR: https://github.com/pytorch/pytorch/pull/72312.
